### ISSUE TITLE
no warning for unused std::error_code

### ISSUE THREAD

**Issue by Trass3r** [llvm/llvm-project#58406]
```cpp
#include <system_error>
void foo()
{
    std::error_code ec;
}
```
The code does get optimized away as system_category() is [[gnu::const]] but there is no -Wunused warning.

https://godbolt.org/z/M68ToY78b

**Comment by adi-g15** [comment]
This is **not** specific to `std::error_code`.

Based on my observation, **any class that has a user defined constructor will _not_ give the warning for `unused`**.

A simpler test case:

```cpp
class Some
{
public:
  // Uncommenting/Removing_all_constructors will give the -unused-variable warning
  // Some() = default;

  // Uncommenting this will NOT give the -unused-variable warning
  Some() {}
};

void foo()
{
    Some a;
}
```

I can't tell Why ? Would be glad if someone else can tell if classes with constructors are treated differently.

The definition for the `-Wunused-variable` flag _for gcc_ is given on [this page](https://gcc.gnu.org/onlinedocs/gcc-12.2.0/gcc/Warning-Options.html#Warning-Options). Quoting that here:

> Warn whenever a local or static variable is unused aside from its declaration. This option implies -Wunused-const-variable=1 for C, but not for C++. This warning is enabled by -Wall.

**Comment by rymiel** [comment]
> Would be glad if someone else can tell if classes with constructors are treated differently.

The rationale is probably that the constructor itself could have side effect, i.e. an RAII resource, and that removing it because it's unused could have side effect. I think from experience GCC performs more checks here.

**Comment by adi-g15** [comment]
Thanks @rymiel, that makes sense 👍.
